Clarified the behavior of the Two-Stage Fallback (#8444)
RasaHQ/rasa: docs/docs/fallback-handoff.mdx

  - active_loop: action_two_stage_fallback
```

> ### 4. Defining an ultimate fallback action (optional)

Reviewer: ```suggestion
### 4. Defining an ultimate fallback action
```

I would not mark this optional because a silent ending to fallback isn't the intended default behaviour

To define the bot's response when the user denies the rephrased intent, define the response `utter_default`:

```yaml-rasa title="domain.yml"
responses:
  utter_default:
  - text: I'm sorry, I can't help you.
```

Or, you can customize `action_default_fallback` for more complex behavior by writing a [Custom Action](./actions.mdx#custom-actions). 
For example, the following snippet is an implementation of a custom action which does the same as `action_default_fallback` but 
dispatches a different template `utter_fallback_template`:

```python title="actions.py"
from typing import Any, Dict, List, Text

from rasa_sdk import Action, Tracker
from rasa_sdk.events import UserUtteranceReverted
from rasa_sdk.executor import CollectingDispatcher

class ActionDefaultFallback(Action):
    def name(self) -> Text:
        return "action_default_fallback"

    def run(
        self,
        dispatcher: CollectingDispatcher,
        tracker: Tracker,
        domain: Dict[Text, Any],
    ) -> List[Dict[Text, Any]]:

        dispatcher.utter_message(template="my_custom_fallback_template")

        # Revert user message which led to fallback.

>         return [UserUtteranceReverted()]

Reviewer: As I rediscovered today, keeping this `UserUtteranceReverted()` in the return events is crucial to not ending up with repeated fallback actions leading to the bot not responding - I forget what it's called, but when it endlessly predicts the same action.

If you can verify this behaviour, we should document it clearly so people don't overwrite the action and get unexpected behaviour.

Reviewer 2: Sorry should have clarified: If you **do not** return `UserUtteranceReverted()` ,  you will see this "catastrophic backtracking"

Reviewer: Sorry should have clarified: If you **do not** return `UserUtteranceReverted()` ,  you will see this "catastrophic backtracking"

Reviewer: Maybe it was something specific to the way I tried it 🤷‍♀️

Author: yes, removing the TED Policy was the root cause.

Reviewer: Right, we do still need overwriting behaviour to work in the absence of TEDPolicy though, so root-root cause is something more.

Reviewer: I think with your explanation below we can resolve this at any rate

> ```

Reviewer: Definitely good to give an example of overwriting.
To indicate why you'd want to overwrite this, the example could be e.g. handoff to human - not implemented, just a mock call to show that's the intention, since that's a very common final fallback

For example, if you want the bot to call a human and stop interacting with the user:

```python title="actions.py"
from typing import Any, Dict, List, Text

from rasa_sdk import Action, Tracker
from rasa_sdk.events import UserUtteranceReverted
from rasa_sdk.executor import CollectingDispatcher

class ActionDefaultFallback(Action):
    def name(self) -> Text:
        return "action_default_fallback"

    def run(
        self,
        dispatcher: CollectingDispatcher,
        tracker: Tracker,
        domain: Dict[Text, Any],
    ) -> List[Dict[Text, Any]]:

        # tell the user they are being passed to a customer service agent
        dispatcher.utter_message(text="I am passing you to a human...")
        
        # assume there's a function to call customer service
        # pass the tracker so that the agent has a record of the converstation between the user
        # and the bot for context
        call_customer_service(tracker)
     
        # pause the tracker so that the bot stops responding to user input
        return [ConversationPaused()]
```

> 

Author: @melindaloubser1 what do you think of this mock handoff?

Reviewer 2: That's great!

    ) -> List[Dict[Text, Any]]:

        # tell the user they are being passed to a customer service agent
        dispatcher.utter_message(text="I am passing you to a human...")
        
        # assume there's a function to call customer service
        # pass the tracker so that the agent has a record of the converstation between the user
        # and the bot for context
        call_customer_service(tracker)
     
        # pause the tracker so that the bot stops responding to user input
        return [ConversationPaused(), UserUtteranceReverted()]
```

  :::caution Events Returned By A Custom Ultimate Fallback Action
    It is a good practice to include `UserUtteranceReverted()` as one of the events returned by your custom
    `action_default_fallback`. Not including this event will cause the tracker to include all events that happened
    during the Two-Stage Fallback process which could intefere with subsequent action predictions from the bot's policy
    pipeline. It is better to treat events that occured during the Two-Stage Fallback process as if they did not happen
    so that your bot can apply its rules or memorised stories to correctly predict the next action. 

>   :::

Author: @melindaloubser1 regarding your comment on on `UserUtteranceReverted()`, what do you think of this update?

Reviewer: That sounds good to me
```suggestion
  :::caution Events Returned By A Custom Ultimate Fallback Action
    You should include `UserUtteranceReverted()` as one of the events returned by your custom
    `action_default_fallback`. Not including this event will cause the tracker to include all events that happened
    during the Two-Stage Fallback process which could intefere with subsequent action predictions from the bot's policy
    pipeline. It is better to treat events that occured during the Two-Stage Fallback process as if they did not happen
    so that your bot can apply its rules or memorised stories to correctly predict the next action. 
  :::
```

Sounds good! Just one language update since we have a "kill list" for writing docs  =)
